nRF5340 nrf-gpio-forwarder

jciupis opened the pull request:
This PR introduces a generic devicetree node that's responsible for forwarding GPIO pins from the nRF5340 application core to the network core. It follows the idea laid out in review of https://github.com/nrfconnect/sdk-nrf/pull/6706

Commit:
dts: add binding for nrf-gpio-forwarder

In nRF53 family of SoCs, GPIO pins must be explicitly forwarded by the
application core to the network core if the latter should drive them.
Add a binding of a generic GPIO pin forwarder.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

Commit:
boards: arm: nordic: update DTS for nRF5340

Add a generic GPIO pin forwarder node to nRF5340 application core DTS
and to Thingy53 application core DTS. For Thingy53, add data of
FEM-related pins to the GPIO forwarder node. For nRF5340, leave the node
with status disabled by default.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

Commit:
soc: arm: nordic: extend nRF5340 initialization

Extend initialization code for nRF5340 application core to iterate over
child nodes of GPIO pin forwarder node and pass control over detected
pins to the network core.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

jciupis commented:
@hubertmis @ankuns please have a look at this draft and share your thoughts on it

hubertmis commented on boards/arm/thingy53_nrf5340/thingy53_nrf5340_common.dts:
```diff
@@ -89,9 +89,24 @@
 		mode-gpios  = <&gpio1 12 GPIO_ACTIVE_HIGH>;
 		pdn-gpios   = <&gpio1 10 GPIO_ACTIVE_HIGH>;
 		tx-en-gpios = <&gpio0 30 GPIO_ACTIVE_HIGH>;
+		gpio-fwd = <&gpio_fwd>;
```
What is the reason to link fem node to nrf-gpio-forwarder?

jciupis replied:
At first I thought that it would be easier to follow this way, but it's unnecessary and turns out to be confusing. I'll remove the reference to `gpio-fwd` from FEM bindings

hubertmis commented on boards/arm/thingy53_nrf5340/thingy53_nrf5340_common.dts:
```diff
@@ -89,9 +89,24 @@
+		compatible = "nordic,nrf-gpio-forwarder";
+		status = "okay";
+		fem-gpio-if {
+			gpios = <&gpio1 11 GPIO_ACTIVE_HIGH>,
```
I understand GPIO flags from these nodes are ignored. I would change their values to 0 and state in the binding file that GPIO flags are ignored, and must be set in the matching binding in the forwarded core device tree nodes using these pins.

hubertmis reviewed (commented):
I like this approach a lot!
Marked a few minor comments to discuss

hubertmis commented on dts/bindings/gpio/nordic,nrf-gpio-forwarder.yaml:
```diff
@@ -0,0 +1,55 @@
+    Here is an example of how a nrf-gpio-forwarder can be used with a nRF21540
+    Front-End module.
+
+    nrf_radio_fem: nrf21540 {
```
I would clarify that `nrf21540` node is required only in DT for Net Core, while `nrf-gpio-forwarder` is required only in DT for App Core. Now user can have an impression that both entries are required for both cores.

hubertmis commented on boards/arm/thingy53_nrf5340/thingy53_nrf5340_common.dts:
```diff
@@ -83,13 +83,18 @@
 		regulator-boot-on;
 	};
 
-	nrf_radio_fem: fem {
```
What happened to the `fem` node? I'm afraid with this patch nRF21540 in Thingy53 might be not working anymore.

hubertmis replied:
Disregard. I see this one was a duplicate of the `fem` node from thingy53_nrf5340_cpunet.dts

ankuns commented on dts/bindings/gpio/nordic,nrf-gpio-forwarder.yaml:
```diff
@@ -0,0 +1,59 @@
+    are ignored. In order to configure the GPIOs passed to the forwarder, their
+    GPIO flags must be set in the matching node that these GPIOs are forwarded
+    to. In the above example, the GPIO flags must be set in the nrf21540 node.
+    They are set to 0 in the nrf-gpio-forwarder node as they are ignored anyway.
```
You provide quite detailed doc. Could you please provide additional info on how to forward pins if there are more net-core nodes as `nrf_radio_fem` having gpio to be forwarded from app core, and these nodes do not know of each other?  
Are there any requirements on naming nodes within `nrf-gpio-forwarder` ?

ankuns replied:
Ok, so please document this fact.

jciupis replied:
@ankuns I added a section to the binding doc.

ankuns commented on boards/arm/nrf5340dk_nrf5340/nrf5340_cpuapp_common.dts:
```diff
@@ -101,6 +101,11 @@
 
+	gpio_fwd: nrf-gpio-forwarder {
+		compatible = "nordic,nrf-gpio-forwarder";
+		status = "disabled";
```
This is by default `"disabled"`. Will each module using forwarder which do not know about other modules using the forwarder need to set it to `"okay"`? Will it result in any conflict?

You provide in this PR update for thingy in file `boards/arm/thingy53_nrf5340/thingy53_nrf5340_common.dts` .
Can the two parts mentioned there `fem-gpio-if` and `fem-spi-if` come for example from two overlays which do not know about each other? What about status `"okay"`? Which one is responsible for setting it to `"okay"` ?

jciupis replied:
You asked a couple of questions, so let me answer them one by one.
> Will each module using forwarder which do not know about other modules using the forwarder need to set it to `"okay"`?

Assuming that the board's definition itself doesn't set it to `"okay"` as e.g. Thingy53 does - yes, that's the safest approach.

> Will it result in any conflict?

As far as I can tell, no.

> Can the two parts mentioned there `fem-gpio-if` and `fem-spi-if` come for example from two overlays which do not know about each other?

Sure, actually I think it's going to be a common use case. Consider an nRF53 application that wants to use both a Front-End module and a Coexistence interface, and have the network core control them. The intended way to handle such a scenario using the solution from this PR would be for the Front-End module overlay to define its own child node of the `nrf-gpio-forwarder` and for the Coexistence interface overlay to define its own. These overlays might be some predefined shields, so in general they don't know about each other. When processing devicetree files for such an application, both these child nodes would end up in a single instance of `nrf-gpio-forwarder` and the initialization code will iterate over them both.

The only requirement for that to work is for both overlays to define child nodes with different names.

> What about status "okay"? Which one is responsible for setting it to "okay" ?

The way I imagined it is that every overlay that needs `nrf-gpio-forwarder` sets its status to `"okay"`.

jciupis replied:
One more comment: I left the `nrf-gpio-forwarder` with status `"disabled"` by default because it seemed cleaner to me to have it disabled if it's unused. However, setting it to `"okay"` by default would simplify the overlays (the status wouldn't have to be modified at all) and decrease the chance of the application getting it wrong. Both solutions are fine IMO.

jciupis replied:
I set the default status to `"okay"`, the solution seems easier to use this way.

hubertmis replied:
Usually for DT entries defined at SOC level status is set to `disabled` (it lists all possible UART, SPI, I2C instances). DT for board enables hardware used in given board (the UART instance connected to external module on the board).

Because `nrf-gpio-forwarder` is defined for board, I think it's OK to set it by default to `okay`.

ankuns reviewed (commented):
Looks very promising, please dispel doubts below

anangl commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -120,6 +130,19 @@ static int nordicsemi_nrf53_init(const struct device *arg)
+	};
+
+	for (int i = 0; i < ARRAY_SIZE(pf_gpios); i++) {
+		soc_secure_gpio_pin_mcu_select(pf_gpios[i].pin,
```
This will work only for `gpio0`, since this function needs a PSEL value (a pin number combined with the related GPIO port number on bit 5) and `pf_gpios[i].pin` is a pin number within a given GPIO port.
You'd rather need to use an array of PSEL values. I think that `soc_nrf_common.h` should be extended with a macro similar to [`NRF_DT_GPIOS_TO_PSEL()`](https://github.com/zephyrproject-rtos/zephyr/blob/52be66919ec454e0943292be325b1fdb8f3ef2ec/soc/arm/nordic_nrf/common/soc_nrf_common.h#L129-L131) but taking additionally the index parameter, and then this macro should be used above instead of `GPIO_DT_SPEC_GET_BY_IDX()`.

jciupis replied:
I added `NRF_DT_GPIOS_TO_PSEL_BY_IDX` as you suggested, please have a look

ppryga commented:
I've got a follow up PR: 
that uses the functionality to forward GPIO pins for Direction Finding Extension of Radio peripheral.
I've tested it and besides comment from @anangl https://github.com/zephyrproject-rtos/zephyr/pull/42576/files#r804800266 it works nicely.

Commit:
soc: arm: nordic: add NRF_DT_GPIOS_TO_PSEL_BY_IDX

Add a macro that allows to retrieve pin's PSEL from its devicetree
representation taking into account its index into GPIO array.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

jciupis commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -46,6 +48,14 @@ extern void z_arm_nmi_init(void);
+#define SPEC_BY_IDX(node_id, prop, idx) \
+	GPIO_DT_SPEC_GET_BY_IDX(node_id, prop, idx),
+#define ALL_GPIOS_IN_NODE(node_id) \
+	(struct gpio_dt_spec) DT_FOREACH_PROP_ELEM(node_id, gpios, SPEC_BY_IDX)
```
With the suggestion from https://github.com/zephyrproject-rtos/zephyr/pull/42576#discussion_r804800266 in place, the type is no longer `gpio_dt_spec`, so the casting has been removed anyway :)

anangl commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -120,6 +129,18 @@ static int nordicsemi_nrf53_init(const struct device *arg)
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP) && \
+	DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_gpio_forwarder)
+	static const int pf_gpios[] = {
```
`uint8_t` should be sufficient. And perhaps `pf_psels` would be a more adequate name for the array.

jciupis replied:
Done.

carlescufi reviewed (changes requested):
Blocking until @mbolivar-nordic has had a chance to review

mbolivar-nordic commented on dts/bindings/gpio/nordic,nrf-gpio-forwarder.yaml:
```diff
@@ -0,0 +1,64 @@
+      };
+    };
+
+    In the above example, the nrf-gpio-forwarder node is configured to forward
+    control over the following GPIO pins:
```
Forward control from where to where? What happens when control is forwarded?

mbolivar-nordic replied:
(To clarify, I'm asking both for myself and because I think this information is missing from the description)

jciupis replied:
I extended the documentation to address your comments, please check if contains all the answers now :)

SebastianBoe commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -120,6 +129,18 @@ static int nordicsemi_nrf53_init(const struct device *arg)
 #endif
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP) && \
+	DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_gpio_forwarder)
```
Let's put this expression in a define to avoid repeating ourselves.

jciupis replied:
Added `NRF_GPIO_FORWARDER_FOR_NRF5340_CPUAPP_ENABLED` to address this comment

SebastianBoe commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -120,6 +129,18 @@ static int nordicsemi_nrf53_init(const struct device *arg)
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP) && \
+	DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_gpio_forwarder)
+	static const uint8_t pf_psels[] = {
```
Remove static. I don't see the need to retain this state between function calls.

ankuns replied:
`static const` is intentional, please have a look on https://github.com/zephyrproject-rtos/zephyr/pull/42576#discussion_r802754077 and following comment https://github.com/zephyrproject-rtos/zephyr/pull/42576#discussion_r802825648

SebastianBoe replied:
Interesting.

So with static it reads the array from flash, but without static it writes the values from flash to the stack, and then reads from the stack.

TIL, thanks.

SebastianBoe commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -120,6 +129,18 @@ static int nordicsemi_nrf53_init(const struct device *arg)
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP) && \
+	DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_gpio_forwarder)
+	static const uint8_t pf_psels[] = {
```
Please use a longer, more descriptive variable name.

jciupis replied:
Replaced with `forwarded_psels`

SebastianBoe commented on soc/arm/nordic_nrf/common/soc_nrf_common.h:
```diff
@@ -120,11 +120,18 @@
+ */
+#define NRF_DT_GPIOS_TO_PSEL_BY_IDX(node_id, prop, idx)	\
+	(DT_GPIO_PIN_BY_IDX(node_id, prop, idx) + \
+	(DT_PROP_BY_PHANDLE_IDX(node_id, prop, idx, port) << 5))
```
NRF_GPIO_PIN_MAP(
  DT_PROP_BY_PHANDLE_IDX(node_id, prop, idx, port),
  DT_GPIO_PIN_BY_IDX(node_id, prop, idx)
  )

the same with the below macro please.

jciupis replied:
Done for both macros

SebastianBoe reviewed (changes requested):
minor changes requested

The approach seems reasonable.

I'm not sure how this relates to non-secure, would need to investigate further.

SebastianBoe commented:
I assume it is not possible for the app core to know of which pins the network core has configured?

There is no mechanism for sharing information between images in DT ?

Because if it did have access to this we could enable pins on the network core normally and then have the app core iterate over those pins and configure them as network pins.

joerchan commented:
> Also let us consider whether this compatible could be used to instruct the secure image to configure pins to be accessible by the non-secure image.

All pins are configured as non-secure. So the opposite is what we want to achieve.
Ideally want we want is bitmask(s) of the pins configured as secure.
https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/master/platform/ext/target/nordic_nrf/common/core/common/tfm_peripherals_config_nrf5340_application.h#L254L264

Can devicetree provide these bitmask(s) to Kconfig?

jciupis commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -45,6 +47,13 @@ extern void z_arm_nmi_init(void);
+#define GPIOS_PSEL_BY_IDX(node_id, prop, idx) \
+	NRF_DT_GPIOS_TO_PSEL_BY_IDX(node_id, prop, idx),
+#define ALL_GPIOS_IN_NODE(node_id) \
+	DT_FOREACH_PROP_ELEM(node_id, gpios, GPIOS_PSEL_BY_IDX)
```
It is necessary - please note the subtle trailing comma at the end

SebastianBoe replied:
ah, I did not see that.

jciupis commented:
> I assume it is not possible for the app core to know of which pins the network core has configured?
> 
> There is no mechanism for sharing information between images in DT ?
> 
> Because if it did have access to this we could enable pins on the network core normally and then have the app core iterate over those pins and configure them as network pins.

That would certainly be a smarter approach as it wouldn't require any duplication in DT files targeted for different cores at all. I'm not aware of any mechanism that would make it possible though.

mbolivar-nordic commented on soc/arm/nordic_nrf/common/soc_nrf_common.h:
```diff
@@ -120,15 +120,25 @@
+/**
+ * @brief Equivalent to NRF_DT_GPIOS_TO_PSEL_BY_IDX(node_id, prop, 0)
+ */
+#define NRF_DT_GPIOS_TO_PSEL(node_id, prop)			\
+	NRF_GPIO_PIN_MAP(					\
+		DT_PROP_BY_PHANDLE(node_id, prop, port),	\
+		DT_GPIO_PIN(node_id, prop))
```
Can you please replace this expansion with `NRF_DT_GPIOS_TO_PSEL_BY_IDX(node_id, prop, 0)` to avoid code duplication?

jciupis replied:
Done

mbolivar-nordic commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -120,6 +134,17 @@ static int nordicsemi_nrf53_init(const struct device *arg)
 
+#if defined(NRF_GPIO_FORWARDER_FOR_NRF5340_CPUAPP_ENABLED)
+	static const uint8_t forwarded_psels[] = {
+		DT_FOREACH_CHILD(DT_INST(0, nordic_nrf_gpio_forwarder), ALL_GPIOS_IN_NODE)
```
I think there is no technical obstacle to allowing the user to create as many of these nodes as they want. Something like this:

```
#define ALL_GPIOS_IN_FORWARDER(node_id) DT_FOREACH_CHILD(node_id, ALL_GPIOS_IN_NODE)

static const uint8 forwarded_psels[] = {
	DT_FOREACH_STATUS_OKAY(nordic_nrf_gpio_forwarder, ALL_GPIOS_IN_FORWARDER)
};
```

That way, the user doesn't have to use `&gpio_fwd { .... };` if they don't want to, and can create their own forwarder nodes as well. That is a bit more flexible and doesn't seem to cost much extra implementation complexity.

Am I missing something?

jciupis replied:
Done

mbolivar-nordic reviewed (commented):
Thank you, this is an excellent description and I understand much better now. A couple of comments.

anangl commented on boards/arm/nrf5340dk_nrf5340/nrf5340_cpuapp_common.dts:
```diff
@@ -101,6 +101,11 @@
 				 <5 &adc 5>;	/* A5 = P0.26 = AIN5 */
 	};
 
+	gpio_fwd: nrf-gpio-forwarder {
+		compatible = "nordic,nrf-gpio-forwarder";
+		status = "okay";
+	};
```
What is the point in defining and enabling the node without specifying any pins to be forwarded (BTW currently, the commit message says "for nRF5340, leave the node with status disabled by default", what is in fact misleading)? I think that either the following pins should be specified here: https://github.com/zephyrproject-rtos/zephyr/blob/d353e551607926612ff258dadbc13d028d58b54e/boards/arm/nrf5340dk_nrf5340/nrf5340_cpunet_reset.c#L16-L20
(and `nrf5340_cpunet_reset.c` updated accordingly) or all changes related to the nRF5340 DK board should be left for a follow-up PR.

jciupis replied:
At first, the node was in `"disabled"` state by default. However, after this comment: https://github.com/zephyrproject-rtos/zephyr/pull/42576#discussion_r802777879 and the discussion that followed, I decided to change the default status to `"okay"`.

I think that the best way to handle the issue is to have the pins you mentioned added to the board's DTS file and leave default status `"okay"`. What do you think?

anangl replied:
> I think that the best way to handle the issue is to have the pins you mentioned added to the board's DTS file and leave default status `"okay"`. What do you think?

If you are willing to do it in this PR, sure, that works for me.

jciupis replied:
Done, please have a look if it's alright now

anangl replied:
Looks good. Thank you!

mbolivar-nordic commented on boards/arm/nrf5340dk_nrf5340/nrf5340_cpunet_reset.c:
```diff
@@ -28,15 +22,6 @@ LOG_MODULE_REGISTER(nrf5340dk_nrf5340_cpuapp, CONFIG_LOG_DEFAULT_LEVEL);
 static void remoteproc_mgr_config(void)
 {
 #if !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE) || defined(CONFIG_BUILD_WITH_TFM)
-	/* UARTE */
-	/* Assign specific GPIOs that will be used to get UARTE from
-	 * nRF5340 Network MCU.
-	 */
-	soc_secure_gpio_pin_mcu_select(CPUNET_UARTE_PIN_TX, NRF_GPIO_PIN_MCUSEL_NETWORK);
-	soc_secure_gpio_pin_mcu_select(CPUNET_UARTE_PIN_RX, NRF_GPIO_PIN_MCUSEL_NETWORK);
-	soc_secure_gpio_pin_mcu_select(CPUNET_UARTE_PIN_RTS, NRF_GPIO_PIN_MCUSEL_NETWORK);
-	soc_secure_gpio_pin_mcu_select(CPUNET_UARTE_PIN_CTS, NRF_GPIO_PIN_MCUSEL_NETWORK);
```
This patch is number 2 in the series, but the calls to soc_secure_gpio_pin_mcu_select() are not added to soc.c until patch number 4. There is therefore a bisectability problem with this PR, but it seems minor and not worth blocking over.

jciupis replied:
I fixed the bisectability problem, the cleaned up commits are present at my fork: https://github.com/jciupis/zephyr/tree/pr/nordic-nrf-gpio-forwarder 

However, apparently there's a bug at Github that might cause a PR to not update when new commits are pushed: https://github.community/t/pull-request-isnt-updating-with-new-commits/226239, and I'm experiencing it now. The latest changes are now shown at all in the PR, commits SHA are not updated, CI is not run, pushes are not reported. I'll wait a couple of hours to see if it's just a temporary glitch, but opening a new PR to replace this one might be necessary.

joerchan replied:
@jciupis Could you link to the new PR?

jciupis replied:
@joerchan https://github.com/zephyrproject-rtos/zephyr/pull/43047

mbolivar-nordic reviewed (approved):
LGTM overall. Ideally the bisectability break would be fixed before merge.

jciupis commented:
I attempted to close the PR and reopen it to work around the issue described in https://github.com/zephyrproject-rtos/zephyr/pull/42576#discussion_r811666998, but Github doesn't allow me to reopen the PR now because `the branch was force-pushed or recreated` (_duh_). Therefore development will continue in #43047, please comment and review there.
